core/preconditioner/isai.cpp

```diff
@@ -205,6 +205,34 @@ void Isai<IsaiType, ValueType, IndexType>::generate_inverse(
     approximate_inverse_ = std::move(inverted);
 }
 
+
+template <isai_type IsaiType, typename ValueType, typename IndexType>
+std::unique_ptr<LinOp> Isai<IsaiType, ValueType, IndexType>::transpose() const
+{
+    std::unique_ptr<transposed_type> transp{
+        new transposed_type{this->get_executor()}};
+    transp->set_size(gko::transpose(this->get_size()));
+    transp->approximate_inverse_ =
+        share(as<Csr>(this->get_approximate_inverse()->transpose()));
```

Comment:
I chose this to avoid the alternative `std::shared_ptr<Csr>(static_cast<...*>(this->get_approximate_inverse()->transpose.release()))`, yes. I personally don't expect a few additional dynamic_casts in core code to have any noticeable runtime impact, especially since they don't need to traverse the type hierarchy in any way, just change the static type.

So the answer to your questions would be: no, it's not necessary, but IMO it makes for much nicer code.